flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -2682,15 +2682,15 @@ class ArrayExprLowering {
     llvm::SmallVector<mlir::Value> extents;
     for (auto extent : seqTy.getShape())
       extents.push_back(builder.createIntegerConstant(loc, idxTy, extent));
-    auto getArrayBox = [&]() {
-      if constexpr (A::category == Fortran::common::TypeCategory::Character) {
-        auto len = builder.createIntegerConstant(loc, idxTy, x.LEN());
-        return fir::CharArrayBoxValue{addr, len, extents};
-      } else {
-        return fir::ArrayBoxValue{addr, extents};
-      }
-    };
-    auto lambda = genarr(getArrayBox());
+    CC lambda;
+    if (fir::isa_char(seqTy.getEleTy())) {
+      auto charTy = seqTy.getEleTy().cast<fir::CharacterType>();
+      auto len = builder.createIntegerConstant(loc, builder.getI64Type(),
+                                               charTy.getLen());
+      lambda = genarr(fir::CharArrayBoxValue{addr, len, extents});
```

Comment:
@schweitzpgi , is there actually a difference between doing:

```
 lambda = return genarr(...);
 return [=](IterSpace iters) { return lambda(iters); };
```

and

```
  return genarr(...);
```

If not, I guess the later would make the code simpler and remove a (Fortran) compile time indirection in the continuation.

Reply:
The simplified code looks ok to me; thanks.

Reply:
There is no difference (other than the extra nop lambda wrapper).